zrml/orderbook-v1/src/lib.rs

```diff
@@ -75,42 +166,45 @@ mod pallet {
         #[transactional]
         pub fn cancel_order(
             origin: OriginFor<T>,
-            asset: Asset<T::MarketId>,
             order_hash: T::Hash,
         ) -> DispatchResultWithPostInfo {
             let sender = ensure_signed(origin)?;
-            let mut bid = true;
-
-            if let Some(order_data) = Self::order_data(order_hash) {
-                let maker = &order_data.maker;
-                ensure!(sender == *maker, Error::<T>::NotOrderCreator);
-
-                match order_data.side {
-                    OrderSide::Bid => {
-                        let cost = order_data.cost()?;
-                        T::Currency::unreserve(maker, cost);
-                        let mut bids = Self::bids(asset);
-                        remove_item::<T::Hash, _>(&mut bids, order_hash);
-                        <Bids<T>>::insert(asset, bids);
-                    }
-                    OrderSide::Ask => {
-                        T::Shares::unreserve(order_data.asset, maker, order_data.total);
-                        let mut asks = Self::asks(asset);
-                        remove_item::<T::Hash, _>(&mut asks, order_hash);
-                        <Asks<T>>::insert(asset, asks);
-                        bid = false;
-                    }
-                }
 
-                <OrderData<T>>::remove(order_hash);
-            } else {
-                return Err(Error::<T>::OrderDoesNotExist.into());
+            let order_data = <Orders<T>>::get(order_hash).ok_or(Error::<T>::OrderDoesNotExist)?;
+
+            let maker = &order_data.maker;
+            ensure!(sender == *maker, Error::<T>::NotOrderCreator);
+
+            match order_data.side {
+                OrderSide::Bid => {
+                    let cost = order_data
+                        .amount
+                        .checked_mul(&order_data.price)
+                        .ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
```

Comment:
Actually, I guess the ideal way to fix this is not to store the price in the order, but rather use ratio-based calculations and store the required fill amount in the order. Basically, instead of storing the price `.1`, we store that filling the entire order will require `100_000_000_000` ZTG (example above).

The frontend can easily calculate the price as `self.fill_amount / self.amount`. The cost of a partial fill can be calculated as

`partial_fill_amount * self.amount / self.fill_amount`.

Reply:
First of all good catch.

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1099/commits/0af56261e1454ab79d79c64b49f81b2e96044b6b